Reinstate boot counting

r-vdp opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

This PR re-introduces boot counting after #330017 was reverted.

There was a proposal to store information about the ESP entries in a separate state file, written down in #337089. However, in this PR we opted rather to use content hashes to allow for matching generations to their ESP entries without necessitating such a state file.

The idea is that we store boot loader entries as files named `nixos-<hash>.conf` with `<hash>` the content hash of the file.
The `systemd-boot-builder` script will then start by listing the system generations of the system, generate a conf file for each (in memory), and checking whether the file, with name derived from the config, already exists, and has the right content.
Missing files will then be created, and additional files (starting with `nixos-`) will be removed.

_(This work is being sponsored by MSF, aka [Doctors Without Borders](https://www.msf-azg.be/en), to help improve resilience for mission-critical deployments in remote environments. MSF is an international, independent medical humanitarian organisation that provides medical assistance to people affected by conflict, epidemics, disasters, or exclusion from healthcare.)_

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

r-vdp committed:
nixos/systemd-boot-builder: format

r-vdp committed:
nixos/systemd-boot-builder: format

r-vdp committed:
nixos/systemd-boot-builder: re-instate boot counting

Co-Authored-By: Julien Malka <julien@malka.sh>
Co-Authored-By: AkechiShiro <14914796+AkechiShiro@users.noreply.github.com>

r-vdp committed:
nixos/systemd-boot-builder: re-instate boot counting

Co-Authored-By: Julien Malka <julien@malka.sh>
Co-Authored-By: AkechiShiro <14914796+AkechiShiro@users.noreply.github.com>

r-vdp committed:
nixos/systemd-boot-builder: store boot loader configs using content hashing

Co-Authored-By: AkechiShiro <14914796+AkechiShiro@users.noreply.github.com>

r-vdp committed:
nixos/systemd-boot-builder: store boot loader configs using content hashing

Co-Authored-By: AkechiShiro <14914796+AkechiShiro@users.noreply.github.com>

JulienMalka commented:
Great work ! Very happy someone is trying to revive boot counting. 

I did not read the code yet, but how is the sorting of the generations working now if efi files essentially have random name?

r-vdp commented:
The file name is only used as a last resort to sort entries after looking at `sort-key`, `version` (which contains the nixos generation number), and `machine-id` (see https://uapi-group.org/specifications/specs/boot_loader_specification/#sorting).
So in practice for us, this means that we sort first by `sort-key`, and then by the generation number.

For the boot default, we don't rely on changing the `sort-key` of the entry (which is surprising, and also annoying when we use content hashing since the content changes based on whether the entry is set as default or not), and instead we set the default ~~using EFI vars~~ in `loader.conf`.

r-vdp commented:
Ugh, not sure what happened there, I fixed the test now, and all systemd-boot tests now succeeded on my machine.

r-vdp commented:
Fixed the merge conflict.

@JulienMalka would you have time to review this?

r-vdp commented:
See https://github.com/systemd/systemd/issues/40192

This was broken upstream in https://github.com/systemd/systemd/pull/38897, reverting that commit makes the test succeed again.

```nix
systemd.package = pkgs.systemd.overrideAttrs (prevAttrs: {
  patches = (prevAttrs.patches or [ ]) ++ [
    (pkgs.fetchpatch2 {
      url = "https://github.com/systemd/systemd/commit/e67e5145675b1849a554aeab4a32badd7d260f9f.diff";
      hash = "sha256-zo7qbSIdMt4YtT/VObmCODJ6hMd363DOBx9ktAznAOY=";
      revert = true;
    })
  ];
});
```

r-vdp commented:
I updated the code now to use the new `preferred` stanza. This is ready to be merged once the current staging-next, which contains systemd 260, will have landed on master.

JulienMalka commented:
I have reviewed the code of this in depth, the design looks sane to me and the implementation too. I think we can merge this once system 260 lands. I would still would like to test this on real hardware before shipping it, but for this I have to wait systemd 260, correct @r-vdp ?

r-vdp commented:
> I have reviewed the code of this in depth, the design looks sane to me and the implementation too. I think we can merge this once system 260 lands. I would still would like to test this on real hardware before shipping it, but for this I have to wait systemd 260, correct @r-vdp ?

Yes, we use a feature from v260 indeed, previous versions would reject the generated loader.conf.
v260 is in staging-next currently, so should land on master soon-ish.

phaer commented:
https://github.com/NixOS/nixpkgs/pull/501049 is on `master` as of ~2h ago and I think this one is good to merge!? Excited to see this coming (back) to NixOS!

r-vdp committed:
nixos/tests/systemd-boot: use a valid machine-id

dbus-broker (now the default since #512050) calls
sd_id128_get_machine() which returns -ENOMEDIUM for an
all-zero machine-id, causing it to crash-loop and the test
to hang on multi-user.target.

r-vdp committed:
nixos/tests/systemd-boot: use a valid machine-id

dbus-broker (now the default since #512050) calls
sd_id128_get_machine() which returns -ENOMEDIUM for an
all-zero machine-id, causing it to crash-loop and the test
to hang on multi-user.target.

r-vdp commented:
Rebased on top of master. The test now succeeds.

@JulienMalka would you be able to do a final check from your side? I think we are good to merge then.

JulienMalka reviewed (approved):
I ran the tests and tested it on my laptop, this is good to go imo.

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -211,19 +315,40 @@ def write_entry(
+    build_time = int(system_dir(profile, generation, specialisation).stat().st_ctime)
+    build_date = datetime.datetime.fromtimestamp(build_time).strftime("%F")
+
+    counters = f"+{BOOT_COUNTING_TRIES}" if BOOT_COUNTING else ""
```
nitpick: Shouldn't `""` be `None`? I think it doesn't create a bug right now, because the logic conditioning on whether it's set will treat it as falsey whether it's `""` or `None`; just seems semantically wrong

r-vdp replied:
Yeah, I guess semantically `None` is better, since the type explicitly allows for it. In the end, we convert `None` to the empty string though, so the result is the same.

I'll push a commit to fix it later tonight.

r-vdp replied:
Done

ElvishJerricco reviewed (commented):
One tiny nitpick, otherwise looks very good to me. Very glad to see this done again, and *especially* grateful you figured out / upstreamed that `preferred` concept for `loader.conf`. That's truly wonderful. Thanks!

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -380,31 +511,67 @@ def install_bootloader(args: argparse.Namespace) -> None:
+    # Garbage-collect stale kernels/initrds/entries before re-populating extra
+    # files, so that user-supplied extraEntries (which may also live under
+    # loader/entries and start with `nixos-`) are not removed again.
+    garbage_collect(gc_roots)
```
Actually, I just realized, this is going to regress on `ENOSPC`. Previously, we would delete the obsolete generations' files before adding new ones, which meant that if `/boot` was full enough that the new files wouldn't fit, room would be made beforehand. By adding the files in `write_entry` and removing them later in `garbage_collect`, we're going to fill up before we make room, right?

Granted, technically, there was a different issue with the old code: if the system nix profile somehow got out of sync with the boot partition (by managing it outside of nixos-rebuild or something) such that all the generations on the boot partition have been removed from the profile, then the old code would have deleted every generation from the boot partition before creating new ones, leaving a brief moment where there is no bootable entry and a crash would leave the system unbootable. The code you have here now doesn't have this problem, since the new entries are always added to the boot partition first. But I think this scenario is substantially more niche than `ENOSPC`.

There's at least one more problem with space, unrelated to this PR, where there has to be space for the `bootctl update` to work. I'm not really sure how we ought to be handling all of these things though. It's tempting to say that we should leave all of this broken in the same ways that it has been for a long time, rather than regressing in the most common fail case (`ENOSPC` when copying boot entries). But I'm really not sure. Maybe the regression is ok since we need a more comprehensive fix anyway?

ElvishJerricco replied:
Arguably the issue I mentioned about the profile going out of sync with the old code isn't a valid issue, since it can't happen as long as `nixos-rebuild` is the only way you add new generations (because in that case the set of generations in the profile can only ever shrink compared to what's on the boot partition, meaning at least one of the entries on the boot partition will point to whatever is left; the issue I described only happens after adding new generations to the profile without `nixos-rebuild`'s behavior of running `stc boot/switch` in all such cases). In that perspective, the new behavior isn't even fixing something else, which makes it just a regression.

We could get it back to how it was before by doing mostly the same code for calculating what *will be* a `GcRoot`, but moving the actual file copying logic to a separate pass after the garbage collection.

What do you all think? Is this a regression that needs fixing?

r-vdp replied:
This sent me down a bit of a spiral, but a fix is coming up. Thanks for noticing this!

r-vdp replied:
Fixed this now, as well as a separate issue with how we were handling initrd secrets that I noticed while working on the first commit.

ElvishJerricco committed:
nixos/systemd-boot: Separate finding the placement of files from writing files

ElvishJerricco committed:
nixos/systemd-boot: Separate finding the placement of files from writing files

ElvishJerricco committed:
nixos/systemd-boot: Rerun secrets every switch

ElvishJerricco committed:
nixos/systemd-boot: Rerun secrets every switch

r-vdp committed:
nixos/systemd-boot-builder: restore EINVAL handling around bootspec loading

#114552 triggers during get_bootspec(), not in write_boot_files(), so
the handler there never fires. Catch it around the generation loop
again so a single broken generation is skipped.

r-vdp committed:
nixos/systemd-boot-builder: write loader.conf after the entries it points at

A crash between the two would leave `default <id>` referring to a
.conf that does not exist yet.

r-vdp committed:
nixos/systemd-boot-builder: write loader.conf after the entries it points at

A crash between the two would leave `default <id>` referring to a
.conf that does not exist yet.

r-vdp committed:
nixos/systemd-boot-builder: rebuild secret-bearing initrds atomically each run

Appending to the existing file made it grow on every rebuild and a
failed script could leave it half-written. Always rebuild from the
pristine initrd into a temp file and rename into place.

r-vdp committed:
nixos/systemd-boot-builder: rebuild secret-bearing initrds atomically each run

Appending to the existing file made it grow on every rebuild and a
failed script could leave it half-written. Always rebuild from the
pristine initrd into a temp file and rename into place.

r-vdp committed:
nixos/systemd-boot-builder: warn instead of aborting when an old gen's secrets fail

After removing or renaming a file in boot.initrd.secrets, older
generations' append scripts start failing. Aborting on that blocks
deploying the new configuration, so only treat a failure as fatal when
it belongs to the configuration being switched to.

r-vdp committed:
nixos/systemd-boot-builder: warn instead of aborting when an old gen's secrets fail

After removing or renaming a file in boot.initrd.secrets, older
generations' append scripts start failing. Aborting on that blocks
deploying the new configuration, so only treat a failure as fatal when
it belongs to the configuration being switched to.

r-vdp committed:
nixos/systemd-boot-builder: use a set for GC root lookup

has_gc_root() iterated the entire BootFileList for every file on the
ESP, giving O(files * roots) comparisons. Build the set of kept paths
once and use O(1) membership tests instead.

r-vdp committed:
nixos/systemd-boot-builder: use a set for GC root lookup

has_gc_root() iterated the entire BootFileList for every file on the
ESP, giving O(files * roots) comparisons. Build the set of kept paths
once and use O(1) membership tests instead.

r-vdp committed:
nixos/systemd-boot-builder: cache boot_path()

It calls Path.resolve() and is invoked several times per generation for
the same store paths.

r-vdp committed:
nixos/systemd-boot-builder: cache boot_path()

It calls Path.resolve() and is invoked several times per generation for
the same store paths.

r-vdp committed:
nixos/systemd-boot-builder: verify content of existing entry files

A file named nixos-<hash>.conf whose content no longer hashes to <hash>
is corrupt. Skip it so GC removes it and a fresh entry is written.

r-vdp committed:
nixos/systemd-boot-builder: verify content of existing entry files

A file named nixos-<hash>.conf whose content no longer hashes to <hash>
is corrupt. Skip it so GC removes it and a fresh entry is written.

r-vdp committed:
nixos/systemd-boot-builder: fall back to pristine initrd when secrets fail

Otherwise the .conf for that generation references a missing initrd and
the boot entry fails to load.

r-vdp committed:
nixos/systemd-boot-builder: fall back to pristine initrd when secrets fail

Otherwise the .conf for that generation references a missing initrd and
the boot entry fails to load.

r-vdp committed:
nixos/systemd-boot-builder: write each ESP path only once

Shared kernels and initrds appear once per generation in boot_files, so
InitrdWithSecretsWriter rebuilt the same file repeatedly. Prefer the
current configuration's entry so its failures stay fatal.

r-vdp committed:
nixos/systemd-boot-builder: write each ESP path only once

Shared kernels and initrds appear once per generation in boot_files, so
InitrdWithSecretsWriter rebuilt the same file repeatedly. Prefer the
current configuration's entry so its failures stay fatal.

asymmetric commented on nixos/doc/manual/release-notes/rl-2605.section.md:
```diff
@@ -386,6 +384,8 @@ See <https://github.com/NixOS/nixpkgs/issues/481673>.
 
 - `stestrCheckHook` was added: This test hook runs `stestr run`. You can disable tests with `disabledTests` and `disabledTestsRegex`.
 
+- `boot.loader.systemd-boot` gained support for [Automatic Boot Assessment](https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT/) via the new [`boot.loader.systemd-boot.bootCounting`](#opt-boot.loader.systemd-boot.bootCounting.enable) options. As part of this change, boot loader entries on the ESP/XBOOTLDR partition are now named `nixos-<content-hash>.conf` instead of `nixos-generation-<n>.conf`; existi ...
```
I think we could expand on this a bit for people who are not aware of what this does "in plain English", eg:

```suggestion
- `boot.loader.systemd-boot` gained support for [Automatic Boot Assessment](https://systemd.io/AUTOMATIC_BOOT_ASSESSMENT/) via the new [`boot.loader.systemd-boot.bootCounting`](#opt-boot.loader.systemd-boot.bootCounting.enable) option, allowing automatic detection of bad boot loader entries. As part of this change, boot loader entries on the ESP/XBOOTLDR partition are now named `nixos-<content-hash>.conf` instead of `nixos-generation-<n>.conf`; existing entries are migrated automatically on the next `nixos-rebuild boot`/`switch`.
```

phaer replied:
If we add this, "bad NixOS generations" might be better than "bad boot loader entries? It's technically about much more, i.e. all of `boot-complete.target`.

And the plural form in `options` seems accurate, as there's `enable` and `tries`?

r-vdp replied:
Both done.

phaer reviewed (approved):
Tested this on one machine; it reverted to an older generation, when testing a broken closure, correctly 🎉 

26.05 has been branched-off. So if the fix for the ENOSPACE issue in  https://github.com/NixOS/nixpkgs/pull/447173#pullrequestreview-4193200039 is good (@ElvishJerricco?), this might be good for master?

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -421,12 +607,80 @@ def install_bootloader(args: argparse.Namespace) -> None:
+            # Keep the entry bootable by leaving at least a pristine initrd
+            # in place. CopyWriter is a no-op if one already exists.
+            assert isinstance(boot_file.writer, InitrdWithSecretsWriter)
+            CopyWriter(source=boot_file.writer.source).write_boot_file(boot_path)
```
This should be done *in the writer implementation*. If there's any other reason for a `CalledProcessError`, say from another writer implementation or whatever, we wouldn't necessarily want this logic handling it. Of course there's no difference right now, since the only way we'd get a `CalledProcessError` is the one we're checking for here, but in general we should be handling the error close to its source. Plus it'd get rid of the `assert isinstance...`

r-vdp replied:
Good point, reworked it so that it's inside the initrd secrets writer now.

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -421,12 +607,80 @@ def install_bootloader(args: argparse.Namespace) -> None:
+                "note: this is normal after having removed "
+                "or renamed a file in `boot.initrd.secrets`",
+                file=sys.stderr,
+            )
```
The message should probably be more instructive. The user is going to be wondering what they should do about it and what impact it has. Maybe something like:

> Warning: Failed to update secrets in the initrd for an older generation ({boot_file.system_identifier.generation}). Stale secrets will continue to be used by this generation. To silence this warning, restore secrets to their original source locations or delete this generation.

r-vdp replied:
Done!

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -421,12 +607,80 @@ def install_bootloader(args: argparse.Namespace) -> None:
+        if bf.path not in unique or bf.current:
+            unique[bf.path] = bf
+
+    for boot_file in unique.values():
```
I did not realize that python dictionaries preserved their insertion order like this (the order is important to make sure entry files come after their kernels / initrds), but it looks like [they have since 3.7](https://stackoverflow.com/questions/2053021/is-the-order-of-a-python-dictionary-guaranteed-over-iterations).

Anyway the `current` handling here moves a file to a different place in the list, which is maybe not actually good; it means older entry files will be written out before their kernels if they share the kernel with the current generation. To be honest I'm not a fan of the `current` field in `BootFile` anyway; I kinda liked that it's *just a file*, not any representation of generations. It would seem cleaner to me that the code that maintains the list of `BootFile`s would also maintain a separate list of which ones are critical, and then the loop here can just pass along the boolean of whether the file is in that critical list when it calls `write_boot_file`.

r-vdp replied:
Things were working correctly already, because we were only doing order-preserving updates. But the rework to get rid of `current` changed this anyway so we don't do any updates like this anymore, so this is a lot cleaner now.

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -50,64 +57,223 @@ class BootSpec:
+    def write_boot_file(self, path: Path) -> None:
+        # Secrets can change between rebuilds, so always rebuild from the
+        # pristine initrd into a temp file and rename into place.
+        tmp = tempfile.NamedTemporaryFile(
```
This should probably still use a `with` statement. As is, the file might not get closed if there's an exception during `copyfileobj`. (This doesn't really matter; we're short-lived and the kernel will close our files when we exit anyway, but I'd still rather just close it properly)

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -50,64 +57,223 @@ class BootSpec:
+        try:
+            with open(self.source, mode="rb") as source_file:
+                shutil.copyfileobj(source_file, tmp)
+            tmp.close()
```
Why was `flush()` not good enough for this before?

r-vdp replied:
This was because we were not using `with` and so the file had to be closed explicitly, I think the `with` got dropped for a reason while refactoring, but it seems that that reason was not there anymore now. We are using `with` again now, so `flush` is sufficient again.

r-vdp committed:
nixos/systemd-boot-builder: clarify stale initrd secrets warning

Tell the user what actually happens (the old secrets stay in place) and
how to get rid of the warning, instead of just saying it is "normal".

Suggested-by: Will Fancher <elvishjerricco@gmail.com>

r-vdp committed:
nixos/systemd-boot-builder: use `with` for the secrets temp file

This guarantees the descriptor is closed even when copyfileobj raises,
matching the other writer implementations. The append-initrd-secrets
script reopens the file by path, so flush() is enough before invoking
it and the explicit close() is no longer needed.

r-vdp committed:
nixos/systemd-boot-builder: track critical paths separately from BootFile

Whether a write failure must be fatal is a property of the destination
path (is it needed by the configuration we are switching to?), not of
the particular BootFile instance that happened to survive
deduplication. Compute the set of critical paths up front and look it
up in write_boot_files, so the dedup loop no longer needs to pick the
"right" instance and becomes a plain order-preserving seen-set walk.

This leaves BootFile.current unused.

Suggested-by: Will Fancher <elvishjerricco@gmail.com>

r-vdp committed:
nixos/systemd-boot-builder: handle initrd-secrets failure in the writer

The CalledProcessError can only come from the append-initrd-secrets
script, so catching it in the generic write loop and then asserting
on the writer type to reach back into its `source` is the wrong
layer. Move the catch, the pristine-initrd fallback and the warning
into InitrdWithSecretsWriter itself, and pass `critical` through the
writer protocol so it can decide between aborting and falling back.

The writer carries the generation number so the warning can still
name the affected generation. write_boot_files no longer knows
anything about secrets and the isinstance assertion is gone.

Suggested-by: Will Fancher <elvishjerricco@gmail.com>

r-vdp committed:
nixos/systemd-boot-builder: drop unused BootFile.{current,system_identifier}

Both fields are now write-only after the previous two commits, so
remove them. BootFile is back to being just a (path, writer) pair.

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -50,64 +57,231 @@ class BootSpec:
+                tmp.flush()
+                run([self.initrd_secrets, tmp.name])
+                with open(tmp.name, "rb") as f:
+                    os.fsync(f.fileno())
```
This `open` was added in the last iteration because of the `tmp.close()` that has now been removed. So there's no need for the extra open here; you can just do `os.fsync(tmp.fileno())` like before.

r-vdp replied:
Right, I'll change that.

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -50,64 +57,231 @@ class BootSpec:
 class SystemIdentifier(NamedTuple):
     profile: str | None
     generation: int
     specialisation: str | None
```
I guess we can just delete this class now. Doesn't seem to be used as of the latest iteration.

r-vdp replied:
We still use it as the type returned from `get_generations`.

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -50,64 +57,231 @@ class BootSpec:
+                os.unlink(tmp.name)
+                if critical:
+                    print("failed to create initrd secrets!", file=sys.stderr)
+                    sys.exit(1)
```
Probably makes more sense to just print the message and then `raise` here to re-raise the `CalledProcessError`, rather than `sys.exit`? Or, I guess the idea is that this way you don't get a whole traceback error message? (That's reasonable, just want to make sure that's what'll happen here and that that's the intent)

r-vdp replied:
Yeah indeed, it's meant to give a clear error message instead of a stack trace. Since this is an issue on the system and not a bug in the script, the stack trace would not be useful and would just produce noise, making it a lot harder for the user to see the actual error message.

ElvishJerricco commented on nixos/modules/system/boot/loader/systemd-boot/systemd-boot-builder.py:
```diff
@@ -380,31 +533,54 @@ def install_bootloader(args: argparse.Namespace) -> None:
             if is_default:
-                write_loader_conf(*gen)
-        except OSError as e:
-            # See https://github.com/NixOS/nixpkgs/issues/114552
```
Looks like we lost this in this iteration. Though frankly I don't understand what it was doing.

r-vdp replied:
That was deliberate, while refactoring this, I realised that the issue that this was referring to, was about specialisations containing special characters that were not supported by vfat, which caused errors. But now we only write hashes with a know fixed prefix, so there is no direct user input anymore that could cause any unexpected characters in the file names, and so this error should never actually occur anymore. So I decided to just take it out.

arianvp reviewed (approved):
this is *excellent* work. Thanks so much for getting this over the finish line
